packages/babylon/src/index.js

```diff
@@ -67,39 +67,39 @@ function getParserClass(
   pluginsFromOptions: $ReadOnlyArray<string>,
 ): Class<Parser> {
   if (
-    pluginsFromOptions.indexOf("decorators-legacy") >= 0 &&
-    pluginsFromOptions.indexOf("decorators") >= 0
+    hasPlugin(pluginsFromOptions, "decorators") &&
+    hasPlugin(pluginsFromOptions, "decorators-legacy")
   ) {
-    throw new Error("Cannot use decorators and decorators2 plugin together");
+    throw new Error(
+      "Cannot use the decorators and decorators-legacy plugin together",
+    );
   }
 
   // Filter out just the plugins that have an actual mixin associated with them.
-  let pluginList = pluginsFromOptions.filter(
-    p => p === "estree" || p === "flow" || p === "jsx" || p === "typescript",
-  );
+  let pluginList = pluginsFromOptions.filter(plugin => {
+    const p = getPluginName(plugin);
```

Comment:
Hey all, I'm concerned about this. Now that we've defined the ability to pass options to Babylon plugins, ordering becomes critical. What if some other piece of tooling injected a syntax plugin into the plugin list after this one, or before? Which one should take priority? For instance for https://github.com/babel/babel/pull/7934/files#diff-3a8233bcd2766d2c7d87f23f944f7726R12, say the user's framework auto-enables `flow`, but the user wants to enable `all`, which comes first?

So far we've avoided this problem by not having options, and not worrying about duplicates. Now it will matter very much both what order your Babel plugins are in, but also what order they enable Babylon plugins in.

At a minimum, have we verified that the ordering on these is consistent with the ordering of visitors? e.g. If two plugins are enabled with different options, whichever one's visitors run first, is also the one who's Babylon configuration takes affect?

Reply:
Let's continue the discussion over in https://github.com/babel/babel/issues/7984 instead of here.